File: library/src/com/lisb/android/mediashrink/MediaShrinkQueue.java
package com.lisb.android.mediashrink;

import java.io.File;
import java.io.IOException;
import java.util.ArrayDeque;
import java.util.Queue;

import org.jdeferred.Deferred;
import org.jdeferred.Promise;
import org.jdeferred.Promise.State;
import org.jdeferred.impl.DeferredObject;

import android.annotation.SuppressLint;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.util.Log;

public class MediaShrinkQueue {

	private static final String LOG_TAG = MediaShrinkQueue.class
			.getSimpleName();

	private static final String WORKING_DIR = "media-shrink-workspace";
	private static final String WORKING_FILE = "shrinking";

	private final Context context;
	private final Handler handler;
	private final Queue<Request> queue = new ArrayDeque<>();
	private final int width;
	private final int videoBitrate;
	private final int audioBitrate;
	private final long durationLimit;

	private ServiceConnection connection;
	private Messenger sendMessenger;
	private Messenger receiveMessenger;

	public static boolean isSupportedDevice(final Context context) {
		return MediaShrink.isSupportedDevice(context);
	}

	public MediaShrinkQueue(Context context, Handler handler, int width,
			int videoBitrate, int audioBitrate, long durationLimit) {
		this.context = context;
		this.handler = handler;
		this.width = width;
		this.videoBitrate = videoBitrate;
		this.audioBitrate = audioBitrate;
		this.durationLimit = durationLimit;
	}

	public Promise<Void, Exception, Integer> queue(Uri source, File dest) {
		final Deferred<Void, Exception, Integer> deferred = new DeferredObject<>();
		final Request request = new Request(deferred, source, dest);
		handler.post(new Runnable() {
			@Override
			public void run() {
				try {
					bindServiceIfNeed();
				} catch (IOException e) {
					Log.e(LOG_TAG, "fail to bind service.", e);
					request.deferred.reject(e);
					return;
				}

				queue.add(request);
				if (sendMessenger != null) {
					sendRequest(request);

				}
			}
		});
		return deferred;
	}

	private void bindServiceIfNeed() throws IOException {
		if (connection == null) {
			Log.v(LOG_TAG, "bind service.");
			connection = new MediaShrinkServiceConnection();
			final Intent intent = new Intent(context, MediaShrinkService.class);
			intent.putExtra(MediaShrinkService.EXTRA_DEST_FILEPATH,
					getWorkingFile().getAbsolutePath());
			intent.putExtra(MediaShrinkService.EXTRA_WIDTH, width);
			intent.putExtra(MediaShrinkService.EXTRA_VIDEO_BITRATE,
					videoBitrate);
			intent.putExtra(MediaShrinkService.EXTRA_AUDIO_BITRATE,
					audioBitrate);
			intent.putExtra(MediaShrinkService.EXTRA_DURATION_LIMIT,
					durationLimit);
			context.bindService(intent, connection, Context.BIND_AUTO_CREATE);
		}
	}

	private void unbindServiceIfNeed() {
		if (queue.isEmpty()) {
			Log.v(LOG_TAG, "unbind service.");
			context.unbindService(connection);
			connection = null;
			sendMessenger = null;
			receiveMessenger = null;
		}
	}

	@SuppressLint("Assert")
	private boolean sendRequest(final Request r) {
		assert sendMessenger != null;
		assert Looper.myLooper() == handler.getLooper();
		final Message m = Message.obtain(null,
				MediaShrinkService.REQUEST_SHRINK_MSGID);
		final Bundle data = new Bundle();
		data.putParcelable(MediaShrinkService.REQUEST_SHRINK_SOURCE_URI,
				r.source);
		m.setData(data);
		m.replyTo = receiveMessenger;
		try {
			sendMessenger.send(m);
			return true;
		} catch (RemoteException e) {
			Log.e(LOG_TAG, "fail to send request.", e);
			return false;
		}
	}

	private void rebindServiceIfNeed() {
		if (!queue.isEmpty()) {
			try {
				bindServiceIfNeed();
			} catch (IOException e) {
				Log.e(LOG_TAG, "fail to reconnect service.", e);
				final Request[] remains = new Request[queue.size()];
				queue.toArray(remains);
				queue.clear();
				for (Request r : remains) {
					r.deferred.reject(e);
				}
			}
		}
	}

	public static boolean isWorkingDirAvailable(final Context context) {
		return Environment.MEDIA_MOUNTED.equals(Environment
				.getExternalStorageState());
	}

	private File getWorkingFile() throws IOException {
		final File dir = context.getExternalFilesDir(WORKING_DIR);
		dir.mkdirs();
		if (!dir.isDirectory()) {
			throw new IOException("Can not create Directory.");
		}
		return new File(dir, WORKING_FILE);
	}

	private class MediaShrinkServiceConnection implements ServiceConnection {
		@Override
		public void onServiceConnected(final ComponentName name,
				final IBinder service) {
			Log.d(LOG_TAG, "onServiceConnected.");
			handler.post(new Runnable() {
				@Override
				public void run() {
					sendMessenger = new Messenger(service);
					receiveMessenger = new Messenger(new ReceiveResultHandler(
							handler.getLooper()));
					Log.v(LOG_TAG, "send all requests. size:" + queue.size());
					for (Request r : queue) {
						if (!sendRequest(r)) {
							// rebind はイベントに任せる
							break;
						}
					}
				}
			});
		}

		@Override
		public void onServiceDisconnected(ComponentName name) {
			Log.e(LOG_TAG, "onServiceDisconnected.");
			handler.post(new Runnable() {
				@Override
				public void run() {
					final Request request = queue.poll();
					if (request != null
							&& request.deferred.state() == State.PENDING) {
						request.deferred.reject(new RuntimeException(
								"process killed."));
					}

					if (connection != null) {
						// 相手方のサービスが自動で復元してしまうことがあるので明示的に unbind しておく。
						context.unbindService(connection);
						connection = null;
						sendMessenger = null;
						receiveMessenger = null;
					}

					rebindServiceIfNeed();
				}
			});
		}
	}

	private class ReceiveResultHandler extends Handler {

		public ReceiveResultHandler(Looper looper) {
			super(looper);
		}

		@Override
		public void handleMessage(Message msg) {
			switch (msg.what) {
			case MediaShrinkService.RESULT_COMPLETE_MSGID: {
				// 動画圧縮の途中でプロセスがkillされた際にゴミファイルが残らないように
				// 圧縮中は作業ファイルに出力し圧縮完了後に出力先に移動するという手順を取る。
				final Request request = queue.poll();
				try {
					getWorkingFile().renameTo(request.dest);
					request.deferred.resolve(null);
				} catch (IOException e) {
					Log.e(LOG_TAG, "fail to rename temp file to dest file.", e);
					request.deferred.reject(e);
				}
				unbindServiceIfNeed();
				break;
			}
			case MediaShrinkService.RESULT_RECOVERABLE_ERROR_MSGID: {
				final Request request = queue.poll();
				request.deferred
						.reject((Exception) msg
								.getData()
								.getSerializable(
										MediaShrinkService.RESULT_RECOVERABLE_ERROR_EXCEPTION));
				unbindServiceIfNeed();
				break;
			}
			case MediaShrinkService.RESULT_UNRECOVERABLE_ERROR_MSGID: {
				// rebind や request のキューからの除去は onServiceDisconnected に任せる
				final Request request = queue.peek();
				request.deferred
						.reject((Exception) msg
								.getData()
								.getSerializable(
										MediaShrinkService.RESULT_UNRECOVERABLE_ERROR_EXCEPTION));
				unbindServiceIfNeed();
				break;
			}
			case MediaShrinkService.RESULT_PROGRESS_MSGID: {
				final Request request = queue.peek();
				request.deferred.notify(msg.arg1);
				break;
			}
			}
		}
	}

	private static class Request {
		public final Deferred<Void, Exception, Integer> deferred;
		public final Uri source;
		public final File dest;

		public Request(Deferred<Void, Exception, Integer> deferred, Uri source,
				File dest) {
			this.deferred = deferred;
			this.source = source;
			this.dest = dest;
		}

	}
}

File: library/src/main/java/com/lisb/android/mediashrink/MediaShrinkQueue.kt
package com.lisb.android.mediashrink

import android.annotation.SuppressLint
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.net.Uri
import android.os.*
import timber.log.Timber
import java.io.*
import java.util.*
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class MediaShrinkQueue(private val context: Context,
                       private val handler: Handler,
                       private val workspace: File,
                       private val width: Int,
                       private val videoBitrate: Int,
                       private val audioBitrate: Int,
                       private val durationLimit: Long) {

    private val queue: Queue<Request> = ArrayDeque()
    private var connection: ServiceConnection
    private var unbindTask: Runnable

    private var sendMessenger: Messenger? = null
    private var receiveMessenger: Messenger? = null
    private var bound = false
    private var unbindInvoked = false
    private var binder: IBinder? = null

    init {
        require(!workspace.isFile) { "workspace must be directory." }
        connection = MediaShrinkServiceConnection()
        unbindTask = UnbindTask()
        cleanWorkspace()
    }

    /**
     * Return file size when succeed to shrink.
     */
    suspend fun queue(source: Uri, dest: Uri, progress: ((Int) -> Unit)? = null) = suspendCoroutine<Long> { continuation ->
        try {
            val request = Request(continuation, progress, source, dest, createWorkingFile())
            handler.post(object : Runnable {
                override fun run() {
                    if (isUnbinding()) {
                        handler.postDelayed(this, DELAY_RETRY_BIND)
                        return
                    }

                    try {
                        bindService()
                    } catch (e: IOException) {
                        Timber.tag(TAG).e(e, "Failed to bind service.")
                        request.deferred.resumeWithException(e)
                        return
                    }

                    queue.add(request)
                    if (sendMessenger != null) {
                        sendRequest(request)
                    }
                }
            })
        } catch (e: IOException) {
            Timber.tag(TAG).e(e, "Failed to create working file")
            continuation.resumeWithException(e)
        }
    }

    @Throws(IOException::class)
    private fun bindService() {
        Timber.tag(TAG).v("bind service.")
        if (binder?.isBinderAlive == true) {
            Timber.tag(TAG).v("Service bound.")
            return
        }
        val intent = Intent(context, MediaShrinkService::class.java)
        intent.putExtra(MediaShrinkService.EXTRA_WIDTH, width)
        intent.putExtra(MediaShrinkService.EXTRA_VIDEO_BITRATE, videoBitrate)
        intent.putExtra(MediaShrinkService.EXTRA_AUDIO_BITRATE, audioBitrate)
        intent.putExtra(MediaShrinkService.EXTRA_DURATION_LIMIT, durationLimit)
        if (!context.bindService(intent, connection, Context.BIND_AUTO_CREATE)) {
            Timber.tag(TAG).e("bindService return false.")
            throw IOException("Failed to connect to MediaShrinkService.")
        }
        bound = true
    }

    private fun unbindServiceIfQueueIsEmpty() {
        if (bound && queue.isEmpty()) {
            Timber.tag(TAG).v("unbind service.")
            context.unbindService(connection)
            bound = false
            unbindInvoked = true
            sendMessenger = null
            receiveMessenger = null
        }
    }

    private fun unbindServiceIfQueueIsEmptyDelayed() {
        handler.removeCallbacks(unbindTask)
        handler.postDelayed(unbindTask, DELAY_UNBIND)
    }

    @SuppressLint("Assert")
    private fun sendRequest(r: Request): Boolean {
        assert(sendMessenger != null)
        assert(Looper.myLooper() == handler.looper)
        val m = Message.obtain(null, MediaShrinkService.REQUEST_SHRINK_MSGID)
        val data = Bundle()
        data.putParcelable(MediaShrinkService.REQUEST_SHRINK_INPUT_URI, r.source)
        data.putString(MediaShrinkService.REQUEST_SHRINK_OUTPUT_PATH, r.workingFile.absolutePath)
        m.data = data
        m.replyTo = receiveMessenger
        return try {
            sendMessenger!!.send(m)
            true
        } catch (e: RemoteException) {
            Timber.tag(TAG).e(e, "Failed to send request.")
            false
        }
    }

    private fun rebindServiceIfQueueIsNotEmpty() {
        if (!queue.isEmpty()) {
            if (isUnbinding()) {
                handler.postDelayed({ rebindServiceIfQueueIsNotEmpty() }, DELAY_RETRY_BIND)
                return
            }
            try {
                bindService()
            } catch (e: IOException) {
                Timber.tag(TAG).e(e, "Failed to reconnect service.")
                val remains = queue.toTypedArray()
                queue.clear()
                for (r in remains) r.deferred.resumeWithException(e)
            }
        }
    }

    private fun isUnbinding(): Boolean {
        if (!unbindInvoked) return false
        if (binder?.pingBinder() != true) {
            unbindInvoked = false
            binder = null
            return false
        }
        Timber.tag(TAG).v("isUnbinding")
        return true
    }

    @Throws(IOException::class)
    private fun createWorkingFile(): File {
        workspace.mkdirs()
        if (!workspace.isDirectory) {
            throw IOException("Can not create workspace.")
        }
        return File.createTempFile(WORKING_FILE_PREFIX, null, workspace)
    }

    private fun cleanWorkspace() {
        Timber.tag(TAG).d("cleanWorkspace")
        val workspaceFiles = workspace.listFiles() ?: return
        for (file in workspaceFiles) {
            file.delete()
        }
    }

    private inner class MediaShrinkServiceConnection : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            Timber.tag(TAG).d("onServiceConnected.")
            handler.post {
                binder = service
                sendMessenger = Messenger(service)
                receiveMessenger = Messenger(ReceiveResultHandler(handler.looper))
                Timber.tag(TAG).v("send all requests. size:%d", queue.size)
                for (r in queue) {
                    if (!sendRequest(r)) { // rebind はイベントに任せる
                        break
                    }
                }
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {
            Timber.tag(TAG).e("onServiceDisconnected:%s", name)
            handler.post {
                val request = queue.poll()
                request?.deferred?.resumeWithException(RuntimeException("process killed."))
                request?.workingFile?.delete()
                bound = false
                sendMessenger = null
                receiveMessenger = null
                rebindServiceIfQueueIsNotEmpty()
            }
        }
    }

    private inner class ReceiveResultHandler(looper: Looper) : Handler(looper) {
        override fun handleMessage(msg: Message) {
            Timber.tag(TAG).d("ReceiveResultHandler#handleMessage. what:%d", msg.what)
            when (msg.what) {
                MediaShrinkService.RESULT_COMPLETE_MSGID -> {
                    // 動画圧縮の途中でプロセスがkillされた際にゴミファイルが残らないように
                    // 圧縮中は作業ファイルに出力し圧縮完了後に出力先に移動するという手順を取る。
                    val request = queue.poll()!!
                    var inStream: InputStream? = null
                    var outStream: OutputStream? = null
                    try {
                        inStream = FileInputStream(request.workingFile)
                        outStream = context.contentResolver.openOutputStream(request.dest)
                                ?: throw IOException("output stream is null")
                        Utils.copy(inStream, outStream)
                        request.deferred.resume(request.workingFile.length())
                    } catch (e: IOException) {
                        Timber.tag(TAG).e(e, "Failed to rename temp file to dest file.")
                        request.deferred.resumeWithException(e)
                    } finally {
                        Utils.closeSilently(outStream)
                        Utils.closeSilently(inStream)
                    }
                    request.workingFile.delete()
                    unbindServiceIfQueueIsEmptyDelayed()
                }
                MediaShrinkService.RESULT_RECOVERABLE_ERROR_MSGID -> {
                    val request = queue.poll()!!
                    val ex = msg.data.getSerializable(MediaShrinkService.RESULT_RECOVERABLE_ERROR_EXCEPTION)
                            as Exception
                    request.deferred.resumeWithException(ex)
                    request.workingFile.delete()
                    unbindServiceIfQueueIsEmptyDelayed()
                }
                MediaShrinkService.RESULT_UNRECOVERABLE_ERROR_MSGID -> {
                    val request = queue.poll()!!
                    val ex = msg.data.getSerializable(MediaShrinkService.RESULT_UNRECOVERABLE_ERROR_EXCEPTION)
                            as Exception
                    request.deferred.resumeWithException(ex)
                    request.workingFile.delete()
                    unbindServiceIfQueueIsEmpty()
                }
                MediaShrinkService.RESULT_PROGRESS_MSGID -> {
                    val request = queue.peek()!!
                    request.progress?.invoke(msg.arg1)
                }
            }
        }
    }

    private inner class UnbindTask : Runnable {
        override fun run() {
            unbindServiceIfQueueIsEmpty()
        }
    }

    private data class Request(val deferred: Continuation<Long>,
                               val progress: ((Int) -> Unit)?,
                               val source: Uri,
                               val dest: Uri,
                               val workingFile: File)

    companion object {
        private const val TAG = "MediaShrinkQueue"
        private const val WORKING_FILE_PREFIX = "working_"

        private const val DELAY_RETRY_BIND = 1000L
        // delay unbind to reuse service
        private const val DELAY_UNBIND = 10 * 1000L

        fun isSupportedDevice(context: Context): Boolean = MediaShrink.isSupportedDevice(context)
    }
}
